code/modules/mob/living/carbon/carbon_update_icons.dm

```diff
@@ -216,84 +218,104 @@
+	var/list/needs_update = list()
+	var/limb_count_update = FALSE
+	for(var/obj/item/bodypart/limb as anything in bodyparts)
+		limb.update_limb(is_creating = update_limb_data) //Update limb actually doesn't do much, get_limb_icon is the cpu eater.
+		var/old_key = icon_render_keys?[limb.body_zone] //Checks the mob's icon render key list for the bodypart
+		icon_render_keys[limb.body_zone] = (limb.is_husked) ? limb.generate_husk_key().Join() : limb.generate_icon_key().Join() //Generates a key for the current bodypart
+		if(!(icon_render_keys[limb.body_zone] == old_key)) //If the keys match, that means the limb doesn't need to be redrawn
+			needs_update += limb
+
+
+	var/list/missing_bodyparts = get_missing_limbs()
+	if(((dna ? dna.species.max_bodypart_count : 6) - icon_render_keys.len) != missing_bodyparts.len) //Checks to see if the target gained or lost any limbs.
+		limb_count_update = TRUE
+		for(var/missing_limb in missing_bodyparts)
+			icon_render_keys -= missing_limb //Removes dismembered limbs from the key list
+
+	if(!needs_update.len && !limb_count_update)
 		return
 
 	remove_overlay(BODYPARTS_LAYER)
 
-	for(var/X in bodyparts)
-		var/obj/item/bodypart/BP = X
-		BP.update_limb()
-
-	//LOAD ICONS
-	if(limb_icon_cache[icon_render_key])
-		load_limb_from_cache()
-		return
-
 	//GENERATE NEW LIMBS
 	var/list/new_limbs = list()
-	var/draw_features = !HAS_TRAIT(src, TRAIT_INVISIBLE_MAN)
-	for(var/X in bodyparts)
-		var/obj/item/bodypart/BP = X
-		new_limbs += BP.get_limb_icon(draw_external_organs = draw_features)
+	for(var/obj/item/bodypart/limb as anything in bodyparts)
+		if(limb in needs_update) //Checks to see if the limb needs to be redrawn
+			var/bodypart_icon = limb.get_limb_icon()
+			new_limbs += bodypart_icon
+			limb_icon_cache[icon_render_keys[limb.body_zone]] = bodypart_icon //Caches the icon with the bodypart key, as it is new
+		else
+			new_limbs += limb_icon_cache[icon_render_keys[limb.body_zone]] //Pulls existing sprites from the cache
+
 	if(new_limbs.len)
 		overlays_standing[BODYPARTS_LAYER] = new_limbs
-		limb_icon_cache[icon_render_key] = new_limbs
 
 	apply_overlay(BODYPARTS_LAYER)
-	update_damage_overlays()
 
 
 
-/////////////////////
-// Limb Icon Cache //
-/////////////////////
+/////////////////////////
+// Limb Icon Cache 2.0 //
+/////////////////////////
 /*
 	Called from update_body_parts() these procs handle the limb icon cache.
 	the limb icon cache adds an icon_render_key to a human mob, it represents:
-	- skin_tone (if applicable)
-	- gender
-	- limbs (stores as the limb name and whether it is removed/fine, organic/robotic)
+	- Gender, if applicable
+	- The ID of the limb
+	- Draw color, if applicable
 	These procs only store limbs as to increase the number of matching icon_render_keys
 	This cache exists because drawing 6/7 icons for humans constantly is quite a waste
 	See RemieRichards on irc.rizon.net #coderbus (RIP remie :sob:)
 */
+/obj/item/bodypart/proc/generate_icon_key()
```

Comment:
`return .` at the end of this

https://github.com/tgstation/tgstation/blob/master/.github/guides/STANDARDS.md#dot-variable-